### PR TITLE
Correctly handle SMT solver errors in loop acceleration code

### PR DESCRIPTION
**Pull request by SaswatPadhi**
Related to #5944 and #5947.

This PR correctly handles SMT solver failures in loop acceleration routines. Earlier these errors were silently ignored, and were being treated as UNSAT.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] NA ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] NA ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- [ ] NA ~~White-space or formatting changes outside the feature-related changed lines are in commits of their own.~~

### CODE REVIEW

File: src/goto-instrument/accelerate/scratch_program.cpp [hunk 1]
```diff
@@ -71,7 +71,12 @@ bool scratch_programt::check_sat(bool do_slice, guard_managert &guard_manager)
   std::cout << "Finished symex, invoking decision procedure.\n";
 #endif
 
-  return ((*checker)() == decision_proceduret::resultt::D_SATISFIABLE);
+  auto result = (*checker)();
+  INVARIANT(
```
Comment (kroening) [inline src/goto-instrument/accelerate/scratch_program.cpp]: `INVARIANT` seems wrong here -- errors do happen, and should be dealt with by giving a proper error message to the user.

Think of `INVARIANT` as something that you could prove to be true, always.

File: src/goto-instrument/accelerate/scratch_program.h [hunk 2]
```diff
@@ -77,8 +77,8 @@ class scratch_programt:public goto_programt
       symex(mh, symbol_table, equation, options, path_storage, guard_manager),
       satcheck(util_make_unique<satcheckt>(mh)),
       satchecker(ns, *satcheck, mh),
-      z3(ns, "accelerate", "", "", smt2_dect::solvert::Z3, mh),
-      checker(&z3) // checker(&satchecker)
+      solver(ns, "accelerate", "", "", smt2_dect::solvert::CPROVER_SMT2, mh),
```
Comment (kroening) [inline src/goto-instrument/accelerate/scratch_program.h]: Do we know how this change of solver affects performance?

File: src/goto-instrument/accelerate/scratch_program.cpp (hunk 1 above)
Comment (SaswatPadhi) [reply src/goto-instrument/accelerate/scratch_program.cpp]: Thanks!

I am not super familiar with the code base, so I am not sure what other construct I could use here to report an error to the user. The function returns a `bool`, so the return value can't capture 3 possibilities: SAT, UNSAT, ERROR.

@kroening  @thomasspriggs do you have any suggestions?

File: src/goto-instrument/accelerate/scratch_program.h (hunk 2 above)
Comment (SaswatPadhi) [reply src/goto-instrument/accelerate/scratch_program.h]: Please ignore this change. I made a mistake in my local testing. This doesn't actually fix the error. @thomasspriggs made a separate change that fixes the root cause.

File: src/goto-instrument/accelerate/scratch_program.cpp (hunk 1 above)
Comment (SaswatPadhi) [reply src/goto-instrument/accelerate/scratch_program.cpp]: Looking at the call sites for this function, I realized that there is exception handling code around it, so probably the right thing to do here is to `throw` an exception and let the caller handle it properly.

**Comment by SaswatPadhi** [comment]
Build failures would be resolved after fa657ac is merged.

**Comment by SaswatPadhi** [comment]
@kroening @thomasspriggs I have addressed your comments now. Could you please take another look?

Thanks.

**Review by kroening — APPROVED** [review]

File: src/goto-instrument/accelerate/scratch_program.cpp (hunk 1 above)
Comment (SaswatPadhi) [reply src/goto-instrument/accelerate/scratch_program.cpp]: Added a `switch` as per Daniel's suggestion.

File: src/goto-instrument/accelerate/scratch_program.cpp [hunk 3]
```diff
@@ -71,7 +71,21 @@ bool scratch_programt::check_sat(bool do_slice, guard_managert &guard_manager)
   std::cout << "Finished symex, invoking decision procedure.\n";
 #endif
 
-  return ((*checker)() == decision_proceduret::resultt::D_SATISFIABLE);
+  switch((*checker)())
+  {
+  case decision_proceduret::resultt::D_ERROR:
+    throw "error running the SMT solver";
```
Comment (chrisr-diffblue) [inline src/goto-instrument/accelerate/scratch_program.cpp:77]: Is there a structure exception you could throw instead of a raw string?

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**